### ISSUE TITLE
Long-Range Depth Model - OAK-D-LR

### ISSUE THREAD

**Issue by Luxonis-Brandon** [luxonis/depthai-hardware#247]
### Start with the `why`:

While the OAK-D Series 1 and 2 ([here](https://docs.luxonis.com/projects/hardware/en/latest/pages/articles/oak-s2.html)) cover medium depth ranges (from ~20cm to 30 meters), and the coming OAK-D405 series (USB https://github.com/luxonis/depthai-hardware/issues/241, PoE https://github.com/luxonis/depthai-hardware/issues/244) specializes in close-in depth (from 0 to 1 or 2 meters), OAK/DepthAI technology is actually capable of much longer depth sensing.  And the only current way to accomplish this is to use OAK-FFC (e.g. [OAK-FFC-3P](https://docs.luxonis.com/projects/hardware/en/latest/pages/DM1090.html) or [OAK-FFC-4P](https://docs.luxonis.com/projects/hardware/en/latest/pages/DD2090.html) with modular cameras such as the [OV9282; 1MP Global Shutter Grayscale](https://shop.luxonis.com/collections/modular-cameras/products/oak-ffc-ov9282-22-pin), [OV9782; 1MP Global Shutter Color](https://shop.luxonis.com/collections/modular-cameras/products/oak-ffc-ov9782-22-pin), or [AR0234; 2.3 MP Global Shutter Color](https://www.arducam.com/product/arducam-full-hd-color-global-shutter-camera-for-raspberry-pi-2-3mp-ar0234-wide-angle-pivariety-camera-module-b0353/).

This works, but is not "production ready" and dealing with FFC cables is just annoying.  So although it is possible to make a wide-stereo-baseline + narrow-FOV stereo pair using the options above (as shown below) such a setup is larger and harder to integrated than is desirable in many cases.  

![image](https://user-images.githubusercontent.com/32992551/168455162-c30e910f-7ab4-4ed8-84d2-fc3a2fc85147.png)

For longer ranges, using the setups above, we've found that a stereo baseline of 15cm (2x the 7.5cm baseline of the OAK-D Series 2), coupled with variable optics, can cover quite a wide range of depth sensing needs.  And also from this testing, we've found the AR0234 to be quite beneficial for long-range sensing, given its large pixel size (matching the OV9282) while having a higher resolution of 2.3MP, which effectively doubling the maximum depth sensing range compared to the OV9282.

The AR0234 also provides the benefit of supporting both global shutter grayscale (for maximizing low-light performance) and also global shutter color (for native pixel-aligned RGBD).

The desired maximum depth range various quite a bit per application - with some situations requiring 100 meters, others 200 meters, and some 300+ meters.  (The furthers the Luxonis team has tested with the AR0234 is 350 meters.)  Supporting M12-mount lenses in the design enables choosing optics with FOVs (fields of view) corresponding to the required sensing distance.  

### Move to the `how`:

Leverage existing DepthAI ecosystem support for AR0234 to implement a dual-AR0234 M12-mount OAK-D-LR.

### Move to the `what`:

- dual AR0234
- M12-mount
- M12 lock rings (to keep focus/alignment)
- USB3.1
- Aluminum Enclosure

**Comment by ynjiun** [comment]
Welcome OAK-D-LR!

I would suggest to include following few features:

add polarizer filter lenses option: since this long range version primarily is for outdoor application, there might be glare caused by the windshield glass, or raining road surface reflection, etc. Adding polarizer might remove or attenuate the glare significantly. 

keembay: wish this new product using latest 3rd Gen VPU!

Ship with 2 sets of lenses: I would suggest to ship this product with two sets of M12 rectilinear lenses: One set of Narrow HFOV (e.g. around 40 to 50 deg) and one set of Wide HFOV (e.g. around 80 to 120 deg). Also the unit should store manufacturing calibrated matrixes for these two sets of lenses. The shipping configuration for example could be preinstalled Narrow_FOV lenses and the default LensMode.Narrow_FOV, such that the rectifiedLeft/Right will assume to apply the Narrow_FOV stereo calibrated matrixes. If user would like to change the lenses to Wide_FOV, then in programming, after user changed the lenses themselves and then need to set LensMode.Wide_FOV to produce correct rectifiedLeft/Right output.

The reason of requesting to ship two lenses sets is a convenient product configuration for an end user like me. Such that I don't need to go out and do the rectilinear lenses sourcing myself and sometime cannot get a quality rectilinear lenses with low quantity like 2 lenses ; )

Future product roadmap:

PoE: wish will have IP67 enclosure PoE version available soon.

**Comment by Luxonis-Brandon** [comment]
Thanks @ynjiun !

> Add polarizer filter lenses option: since this long range version primarily is for outdoor application, there might be glare caused by the windshield glass, or raining road surface reflection, etc. Adding polarizer might remove or attenuate the glare significantly.

This is the beauty of M12.  This makes it possible to simply buy M12 solutions that have this capability built-in.  

> keembay: wish this new product using latest 3rd Gen VPU!

Great point.  We can likely design this using the OAK-SoM-Pro or OAK-SoM-MAX to have KeemBay support.  

> Ship with 2 sets of lenses: I would suggest to ship this product with two sets of M12 rectilinear lenses: One set of Narrow HFOV (e.g. around 40 to 50 deg) and one set of Wide HFOV (e.g. around 80 to 120 deg). Also the unit should store manufacturing calibrated matrixes for these two sets of lenses. The shipping configuration for example could be preinstalled Narrow_FOV lenses and the default LensMode.Narrow_FOV, such that the rectifiedLeft/Right will assume to apply the Narrow_FOV stereo calibrated matrixes. If user would like to change the lenses to Wide_FOV, then in programming, after user changed the lenses themselves and then need to set LensMode.Wide_FOV to produce correct rectifiedLeft/Right output.

Great suggestion on lens types/etc. and agreed on the FOVs here.  One catch is that once the lenses are removed, the calibration is invalidated.  So probably what is best here is to have just 2 different purchase options as the default for this model, such that they can come pre-calibrated.  And then there's a decent chance that for a given purpose these might be "Good enough".  And then a no-lens option for folks who need their own lenses (so as to not waste lenses/etc.)
1. Narrow FOV (e.g. around 40 to 50 deg), lenses pre-installed and lock-ringed in place (likely with breakable staking), factory calibrated.
2. Wide FOV (e.g. around 80 to 120 deg), lenses pre-installed and lock-ringed in place (likely with breakable staking), factory calibrated.
3. No lenses.  Lock-rings included in gift box.

We'll then also sell lenses on our site, just like our friends OpenMV do, [here](https://openmv.io/collections/lenses).  Note they even have the polarization filter as well, which would very likely work here (not yet tested of course, as we haven't even started work on OAK-D-LR yet).  

> The reason of requesting to ship two lenses sets is a convenient product configuration for an end user like me. Such that I don't need to go out and do the rectilinear lenses sourcing myself and sometime cannot get a quality rectilinear lenses with low quantity like 2 lenses ; )

Agreed.  Good points.  Doing them as separately-orderable options will save cost.  And it's required for factory-calibration to be applied and usable.

> Synced Left/Right requirement: I am not sure the current OAK-D left/right sync requirement in terms of time difference. I would suggest that new product left/right sync frame time difference < 100us. If it can be less than 50us or smaller would be even better. When the left/right exposure start/end time not in sync, the disparity map will not be accurate for moving objects. The problem will be amplified particularly when the ego view is turning (with angular speed): for example, when a car is turning at the cross road, say at 90 deg/ 5.4 sec => 16.6 deg/sec, then at around 25 m away, there will be disparity error by 1 pix for every 1 ms of out of sync left/right exposure. The further the distance, the bigger the disparity error will be (at around 250m, then it will be amplified by 10X..)

Yes.  We'll do hardware sync here.  My understanding of the OAK-D sync is it's likely in the sub-1-microsecond range.  It gets very hard to actually test that though in terms of pixels.  But as far as our testing has revealed (leveraging MIPI readouts/etc.) we're at least under a couple microseconds.  And the same would be true here.  

> PoE: wish will have IP67 enclosure PoE version available soon.

Agreed.  We'll do a PoE as well.  I actually think PoE will be more important here.  USB is a tad quicker to get out to test optics/etc. so we'll simply do that first - to learn the `unknown unknown` here in terms of the discovery that always occurs on a new product like this.

Thanks again!
-Brandon

**Comment by stephansturges** [comment]
I would be very interested in this product as well, especially the PoE version with an IP67 enclosure. 

On the subject of filters and mechanics I would love the option to order the camera with a set of two "front plates", with the glass on one of these plates incorporating a polarization layer directly if that's possible. 

Or alternatively (but more complicated) having a gasket-sealed "hatch" on the side of the assembly that would allow to slide a filter in front of the 3 lenses, held in place by a groove for example. This could then also allow the use of other filters for specific wavelengths of light or ND filters to reduce exposure without mounting anything to the outside of the sensor.... but I realise this second option is more complicated as the use of filters would depend on the size of the lenses and sensors on the board.

**Comment by ynjiun** [comment]
@Luxonis-Brandon 

> Yes. We'll do hardware sync here. My understanding of the OAK-D sync is it's likely in the sub-1-microsecond range. It gets very hard to actually test that though in terms of pixels. But as far as our testing has revealed (leveraging MIPI readouts/etc.) we're at least under a couple microseconds. And the same would be true here.

Question: is that possible to sync all 3 cameras including the center 4K camera with the other two stereo cameras?

Center 4K camera position: for this 15cm baseline version, I would recommend the center 4K camera position at right side with distance to right camera of 5cm and distance to the left camera of 10cm. Plus if possible to sync all three cameras, then we might have a stereo system with 3 baselines: 15cm, 10cm, and 5cm in operation simultaneously. If the software and keembay hardware throughput can handle 3 pairs of stereo disparity computation, then we could have a hybrid system that fuse 3 disparity maps together to generate a much higher accuracy disparity map. What do you think? I would happy to join the prototyping work on this front if you can send me an early prototype ; ))

**Comment by Luxonis-Brandon** [comment]
> On the subject of filters and mechanics I would love the option to order the camera with a set of two "front plates", with the glass on one of these plates incorporating a polarization layer directly if that's possible.

I love this idea!  We will do this for sure.  

And yes, I do like the idea of like separate front covers for this.  So that they can still be sealed.  

Thanks,
Brandon

**Comment by Luxonis-Brandon** [comment]
For this one we were thinking only 2x AR0234 global shutter color 2.3MP.  So no 4K rolling shutter at all.  Thoughts on that?

**Comment by stephansturges** [comment]
For my applications rolling sur if any kind is a no-go, and 2MP global shutter sensors would actually be ideal for balance between light sensitivity and quantity of data :)

*Stephan Sturges*
+32 484 06 19 63

**Comment by gtech888AU** [comment]
Brandon, 
Have you considered adding a design feature which would allow users to be able to modify the camera angles to face inwards rather than being forced to keep them parallel.  This would open up usage of this model across different depth ranges (both near and far) due to its ability to customize the lens into much narrower FOVs if desired - to date, this is something your other cameras besides FFC all lack.

**Comment by Luxonis-Brandon** [comment]
Hi @gtech888AU ,

So actually for stereo disparity depth having cameras angled-in prevents the disparity depth from working properly.  As with disparity depth the key is for the cameras to see the information from the same vantage point, just displaced.  And with angled cameras, one camera will see the left side of an object, and the other will see the right - and so feature matching won't work.

That and any angle between the sensors reduces FOV of the depth.  So it's a disadvantage to do so all around.

Hi @stephansturges,

Thanks, makes sense.  And PS I edited your post to remove private information.

Thanks all,
Brandon

**Comment by ynjiun** [comment]
> For this one we were thinking only 2x AR0234 global shutter color 2.3MP. So no 4K rolling shutter at all. Thoughts on that?

I think no 4K rolling shutter is fine. 

Then I would suggest to enhance this unit with an extra AR0234 global shutter (thus total of 3x AR0234) at the following position:

left <- 1.5 cm -> middle <- 13.5cm -> right

So the left/right baseline is 15cm
And middle/right baseline is 13.5cm
And left/middle baseline is 1.5cm

All three cameras are in h/w sync.

**Comment by Luxonis-Brandon** [comment]
YES!  That's a great idea @ynjiun .  As that gives both long-range and not-long-range depth out of the same model.  

And with that idea I'm now pondering what the "right" distribution of baselines is.  

My gut is something like the following might be super useful.

left <- 5cm -> middle <- 10cm -> right

As this gives the 3 following baselines:
1. 15cm
2. 10cm
3. 5cm

The `why` of this distribution is that the "middle" baseline of 13.5cm and widest baseline of 15cm feel very close in terms of usable range.  And likely are.  Whereas with something like this spread, the "middle" baseline is a bigger jump from the widest baseline, with 10cm covering a more-different depth range than 15cm.  

And then similar to 5cm, that can allow fairly-close in depth compared to both 10cm an 15cm.

Thoughts?

Thanks,
Brandon

**Comment by ynjiun** [comment]
@Luxonis-Brandon 

> left <- 5cm -> middle <- 10cm -> right

Agree. This would definitely address those who need different baseline in a meaningful seperation.

The reason I propose a different set of
> left <- 1.5cm -> middle <- 13.5cm -> right
is inspired by this [paper ](https://www.ri.cmu.edu/pub_files/pub2/okutomi_m_1993_1/okutomi_m_1993_1.pdf)from CMU.
This configuration might provide both "accuracy" and "robustness" extending the 15cm baseline limitation. (I hope ; ))

**Comment by Luxonis-Brandon** [comment]
Fascinating.  Thanks.  I just updated the specs at the very top with the culmination of discussions here.  Added [NEW] for things that came from the discussion.

**Comment by stephansturges** [comment]
I would love to know if there is already a tentative timeline attached to this project?

**Comment by Luxonis-Brandon** [comment]
Unknown as of yet.  We're triaging internally on when we can implement this.  We'll circle back when we have a better view of schedule.

We're also triaging interest on this vs. other models.  So this helps that there's interest.  And if you want to pre-order, we can get you links for that.  :-). Helps with customer voting.  ;-)

**Comment by microaisystems** [comment]
@Luxonis-Brandon 

#And if you want to pre-order, we can get you links for that. :-). Helps with customer voting. ;-)

Yes, send me the pre-order link.

**Comment by Luxonis-Brandon** [comment]
Thanks @microaisystems .  Will get it up now.  :-)

**Comment by Luxonis-Brandon** [comment]
It's live here:
https://shop.luxonis.com/products/oak-d-lr-pre-order

(The LR will have to be quite a bit more expensive because it has the (expensive) 2.3MP global shutter color and (expensive) large/nice optics, which then also make the whole thing bigger (and more expensive).)

**Comment by stephansturges** [comment]
This is like computer vision catnip to me... I've ordered 2 units 😄

**Comment by ynjiun** [comment]
@Luxonis-Brandon 

Great! Before I order d-lr, one question: would this pre-ordered unit shipped with Myriad X or KeemBay? would this unit have Raspberry Pi Module 4 integrated as well? Thank you.

**Comment by Erol444** [comment]
Hi @ynjiun,
We will likely design the OAK-D-LR with the [OAK-SOM-PRO](https://docs.luxonis.com/projects/hardware/en/latest/pages/BW2099.html), so users will **be able to choose either the Myriad X** (BW2099) version **or KeemBay** (DM2399) version:)! Thoughts?
Thanks, Erik

**Comment by stephansturges** [comment]
The option to use either module would be very interesting to me, but especially if they are user-replaceable. 
ie: I really want to get working with the Keem Bay architecture but there are things I would like to test immediately which I know will work on the Movidius version without breaking any of my code. 
So the ideal scenario would be to have the option to "upgrade" to the leem bay by swapping out a module, if at all possible.

**Comment by Erol444** [comment]
Hi @stephansturges,
I believe these would indeed be user replaceable! As we designed KeemBay version of OAK-SOM-PRO so it's **backwards compatible** with the MyriadX version of OAK-SOM-PRO:)
Thanks, Erik

**Comment by stephansturges** [comment]
That's ideal, thanks.

**Comment by stephansturges** [comment]
Hey guys! I see on the product page in the store that you are targeting Jan 2023, is this still correct? If there's any update ton the timeline I'd love to know more :)

**Comment by Luxonis-David** [comment]
Quick update on the project, we have sent the design to FAB house and we are estimating to get back the first prototypes in a month or so. Below are a few renders a sneak peak to what we are working on. 
Please note those renders does not necessarily represent the look of the final product and any changes can be made without prior notice.
![image](https://user-images.githubusercontent.com/75903226/197033358-1ac9fac5-9430-4243-87e0-5d0b2b1cf86c.png)
![image (104)](https://user-images.githubusercontent.com/75903226/197033387-cc3f6d71-fb2f-40b2-a6c9-71404c134060.png)
![image (103)](https://user-images.githubusercontent.com/75903226/197033409-df809ca2-f010-4053-99dc-3875c3d76e0c.png)

**Comment by stephansturges** [comment]
Looking good! I can't wait to use this.

**Comment by chad-green** [comment]
> YES! That's a great idea @ynjiun . As that gives both long-range and not-long-range depth out of the same model.
> 
> And with that idea I'm now pondering what the "right" distribution of baselines is.
> 
> My gut is something like the following might be super useful.
> 
> left <- 5cm -> middle <- 10cm -> right
> 
> As this gives the 3 following baselines:
> 
> 1. 15cm
> 2. 10cm
> 3. 5cm
> 
> The `why` of this distribution is that the "middle" baseline of 13.5cm and widest baseline of 15cm feel very close in terms of usable range. And likely are. Whereas with something like this spread, the "middle" baseline is a bigger jump from the widest baseline, with 10cm covering a more-different depth range than 15cm.
> 
> And then similar to 5cm, that can allow fairly-close in depth compared to both 10cm an 15cm.
> 
> Thoughts?
> 
> Thanks, Brandon

@Luxonis-Brandon the renderings look pretty awesome. What did you guys decide on for the baselines?

**Comment by Luxonis-David** [comment]
@chad-green we used 5cm for shorter and 15cm for longer baseline

**Comment by cafemoloko** [comment]
Here's a sneak peek of the OAK-D-LR enclosure. However, this is not a completely assembled prototype, and the team is working on fine-tuning it.

![image (3)](https://user-images.githubusercontent.com/26538615/201292368-4bf4287d-8ea5-443a-a038-3667e8905e49.png)
![image (2)](https://user-images.githubusercontent.com/26538615/201292395-2cf15d2d-b86b-49cf-b581-5141c4f0b74d.png)
![image](https://user-images.githubusercontent.com/26538615/201292407-fc5dce3a-0dd6-4263-8a26-c5fdd1b7308d.png)

**Comment by stephansturges** [comment]
This made my day @cafemoloko ! I'm so excited to start playing with this one!

**Comment by Luxonis-David** [comment]
Adding 3D CAD file here so one can start tinkering on the integration of the OAK-D-LR into bigger system (design is not final and is a subject to change during development process).
[OAK-D-LR_TopAssy-24NOV2022.zip](https://github.com/luxonis/depthai-hardware/files/10084479/OAK-D-LR_TopAssy-24NOV2022.zip)

**Comment by Alex-Beh** [comment]
May I know where can I find the specifications? Looking forwards to try this amazing product!!!

**Comment by Luxonis-David** [comment]
@Alex-Beh some of the specs are written in the original post of this issue, I am pasting and adding more below:

- Triple AR0234 2.3MP global shutter color sensor (DFOV 96°) 
- Camera Spacing {right} <-5cm-> {middle} <-10cm-> {left} 
- 3 stereo baselines: 5cm, 10cm, and 15cm for short, medium and long-range depth
- Swappable M12-mount lenses with M12 lock rings (to keep focus/alignment)
- Front Gorilla Glass (same as we use on other OAK-D products)
- USB3.1 and 802.3at PoE
- Aluminum Enclosure 
- Rough dimensions: 202 x 43 x 40 mm (without rain cover)
       - Rain cover dimensions: 50 x 27 x 22 mm
- 75mm Vesa mount M4 screws 
- 1/4" tripod mount in center
- Recovery button and RGB status LED on the bottom 
- OAK-SoM-Pro based 
- Not IP sealed
- Comes with the cables rain cover, protecting the device from splashing water
![image](https://user-images.githubusercontent.com/75903226/204464686-2ebafc46-8257-4b02-8429-dceac8bf7658.png)
![image](https://user-images.githubusercontent.com/75903226/204464790-4387fdb2-cd65-4936-9cea-145e3df9bee6.png)
![image](https://user-images.githubusercontent.com/75903226/204464932-f1bcde6c-0ca3-4df8-bc80-9ef82bd86cb9.png)

**Comment by ynjiun** [comment]
@Luxonis-David 

"OAK-SoM-Pro based"

could we ship with OAK-SoM-Max module? is it compatible? I would like to try the new module with OAK-D-LR. Please advise.

Thanks.

**Comment by Luxonis-David** [comment]
@ynjiun no unfortunately it won't be compatible with OAK-SoM-MAX. 
We did on the other hand design OAK-TBot which is compatible with OAK-SoM-MAX, we just released the design for fabrication so there is still a long way (two months) to go before we will have it officially supported but it will be soon in the Early access store so you can reserve one from the first batch.  
It comes with a mount and it features the following:
- Camera sensor module selection:
    - 2x IMX582 120° DFOV (96° HFOV, PY080)
    - IR capable IMX582
    - IR filtered IMX582
    - 3x OV9282 Standard FOV (80° HFOV)
         - 7.5cm H-stereo
         - 2.0cm V-stereo
- 1x IR laser dot projector, w/ or w/out it - the camera should look good (balanced) in both cases
- Multiple IR LEDs (still need to figure out which ones) 
- uSD card slot covered with a silicon rubber, on the bottom side - must support outdoor installs
- USB accessible and covered with a silicon rubber
- PoE on the back with vertical RJ45 connector 
- RVC3 based product 
- Large system LED
- Factory reset button, next to the ethernet connector
- Camera mount included (wall mounting, pole mounting)
- Indoor/Outdoor installation
- Microphones - 2 in front and 1 on the back side

The purpose of this product is that it would cover logistics and light manufacturing. The idea is to have an easy-to-setup and “just works” camera that non-engineers can use for things such as:
- Box measurement
- QR and/or bar-code scanning 
- People counting/trac
- Basic quality/inspection/etc.

See below attached renders: 
![image](https://user-images.githubusercontent.com/75903226/204741077-43d6e240-794d-4e02-a6eb-fb43058e054d.png)

**Comment by Luxonis-David** [comment]
Hi @stephansturges, 

Good question, OAK-D-LR is currently being assembled in one of our offices and we can start shipping in around 14-days. 
We do need to add a support for new "dual" bootloader on as OAK-D-LR exposes both USB and ETH ports, this is also the reason for above ETA.

**Comment by alexivins** [comment]
@Luxonis-David
Hi David, any chance you could send the depth coverage for the OAK-TBot? I am looking to see what type of coverage for placing camera at a height of .3 meter to .5 meters (12 inches to 22 inches). I am assuming based on cameras this has Depth capability. Could you also share details on the mount as well. Also what is the price point for this one? Thank you.

**Comment by ynjiun** [comment]
> * Triple AR0234 2.3MP global shutter color sensor (DFOV 96°)

With this DFOV 96°, what would be the depth range for 15cm baseline pair?

Could we ship with a smaller FOV option, for example: FOV (D/H/V) 50°/40°/30° to get farther depth range?

Thanks.

**Comment by Luxonis-David** [comment]
> @Luxonis-David Hi David, any chance you could send the depth coverage for the OAK-TBot? I am looking to see what type of coverage for placing camera at a height of .3 meter to .5 meters (12 inches to 22 inches). I am assuming based on cameras this has Depth capability. Could you also share details on the mount as well. Also what is the price point for this one? Thank you.

The product is still under the design process and I can't share the details about it at the moment. On OAK-TBot we expect that for a start  the same stereo cameras will be used as on OAK-D-Pro so NFOV OV9282 with bandpass filter for model with active illumination and with cutoff filter for model with no IR illumination camera FOV (D/H/V) | 89.5°/80°/55°, with horizontal baseline 7.5cm and vertical 2.5cm. 
You might be able to calculate the depth coverage based on that IMO but note that anything of the above might be subject of change after we test out the first prototypes and find what would be best. 
About the mount it is also not finalized, we expect having several different mounts, besides the one on the photo above we do plan on having one meant for more "rigid" mounting.
![image](https://user-images.githubusercontent.com/75903226/206110872-9c3b8cc1-550a-47c4-8534-112f3acafa61.png)

I will follow up here when we do have more details about the asked.

**Comment by Luxonis-David** [comment]
> > * Triple AR0234 2.3MP global shutter color sensor (DFOV 96°)
> 
> With this DFOV 96°, what would be the depth range for 15cm baseline pair?
> 
> Could we ship with a smaller FOV option, for example: FOV (D/H/V) 50°/40°/30° to get farther depth range?
> 
> Thanks.

Yes we can, but MOQ and extended lead time would probably apply in order to do custom FOV lens assembly. 
For the lead time in such cases we do normally quote 8 weeks by default, but it depends on if we do already have a batch in production and we can make a custom assembly out of the same batch to shorten the lead time to just few days/weeks.

**Comment by ynjiun** [comment]
Hi @Luxonis-Brandon 

Perhaps it's time to talk about the software support of this new product (OAK-D-LR). The following are few thoughts and questions would like to get your feedback:

Architecture: 3 pairs of disparity computation: LR has 3 pairs of stereo: right/left (baseline=15cm), right/middle (baseline=5cm) and middle/left(baseline=10cm).
Questions:
1. can we support StereoDepth pipeline of 3 pairs at the same time (concurrently)? That is to provide a pipeline API to connect left/middle/right camera to pipe into an API as below:
```
               ┌───────────────────┐
               │                   │ confidenceMap
               │                   ├─────────────►
               │                   │rectifiedLeft
               │                   ├─────────────►
left           │                   │   syncedLeft
──────────────►│-------------------├─────────────►
               │                   │rectifiedMiddle
               │                   ├─────────────►
middle         │                   │   syncedMiddle
──────────────►│-------------------├─────────────►
               │                   │   depth [mm] (right/left, right/middle, middle/left)
               │                   ├─────────────►
               │    StereoDepth    │    disparity (right/left, right/middle, middle/left)
               │                   ├─────────────►
right          │                   │   syncedRight
──────────────►│-------------------├─────────────►
               │                   │rectifiedRight
               │                   ├─────────────►
inputConfig    │                   |     outConfig
──────────────►│-------------------├─────────────►
               └───────────────────┘
```


2. what would be the throughput/bandwidth estimate to support left/middle/right StereoDepth API? can we still have >= 30 fps for 1280x800 resolution?

3. can we still support -dumpdispcost API?

4. is that possible to provide a call-back API to allow user to define a function to fuse 3 dumpdispcost says (3,height,width,96) into 1 dumpdispcost (1,height,width,96) and pipe it back to compute the final depth and disparity? If this is possible, then the above API achitecture can be simplified to a single depth and disparity output instead of 3X each.

Could we have a zoom or separate email discussion on this? 

Please advise. Thanks a lot for your help.

**Comment by Luxonis-David** [comment]
@themarpe would you be able to give some answers to above already?

**Comment by Luxonis-David** [comment]
@alexivins please check on a separate issue for OAK-TBot #326.
We can move further discussion there.

**Comment by ynjiun** [comment]
> One will be able to place 2 StereoDepth nodes to achieve stereo between all 3 cameras at the same time. The performance remains inline with maximum capabilities that Stereo on RVC2 can be done, but its split between these two instances + some overheads.

Hi @themarpe 

Thank you for the information.

Couple more questions:
1. "place 2 StereoDepth nodes": is there a limit on how many StereoDepth nodes we can place? for example, could we place 3 nodes (right/left pair, right/middle pair and middle/left pair) concurrently?
2. would RVC3 provide higher throughputs for StereoDepth? by how much? does it support 1920x1080 resolution? Do we have a RVC3 performance table similar to [RVC2 table](https://docs.luxonis.com/projects/api/en/latest/components/nodes/stereo_depth/#stereo-depth-fps) you provided? 

Thank you for your help.

**Comment by szabi-luxonis** [comment]
> > One will be able to place 2 StereoDepth nodes to achieve stereo between all 3 cameras at the same time. The performance remains inline with maximum capabilities that Stereo on RVC2 can be done, but its split between these two instances + some overheads.
> 
> 
> 
> Hi @themarpe 
> 
> 
> 
> Thank you for the information.
> 
> 
> 
> Couple more questions:
> 
> 1. "place 2 StereoDepth nodes": is there a limit on how many StereoDepth nodes we can place? for example, could we place 3 nodes (right/left pair, right/middle pair and middle/left pair) concurrently?
> 
> 2. would RVC3 provide higher throughputs for StereoDepth? by how much? does it support 1920x1080 resolution? Do we have a RVC3 performance table similar to [RVC2 table](https://docs.luxonis.com/projects/api/en/latest/components/nodes/stereo_depth/#stereo-depth-fps) you provided? 
> 
> 
> 
> Thank you for your help.
> 
> 
> 
> 

1. No implicit limit.
2. Yes, it has higher throughput. Maximum 1280 width is supported. We don't have public performance table.

**Comment by ynjiun** [comment]
> We will likely design the OAK-D-LR with the [OAK-SOM-PRO](https://docs.luxonis.com/projects/hardware/en/latest/pages/BW2099.html), so users will **be able to choose either the Myriad X** (BW2099) version **or KeemBay** (DM2399) version:)! Thoughts?

I already pre-ordered OAK-D-LR. How do I notify Luxonis to ship my order with KeemBay (DM2399)? Please advise. Thanks.

**Comment by Erol444** [comment]
@ynjiun please send an email to support@luxonis.com.

**Comment by Luxonis-David** [comment]
**! Update on OAK-D-LR compatibility with RVC3 based OAK-SoM-Pro-S3 (DM2399) !**

We will be fulfilling first batch of samples assembled with RVC2 OAK-SoM-Pro only.
The reason for being, DepthAI is not yet fully supported on RVC3 and we are also working on missing working drivers for the `PCIe --> ETH` bridge used in the product. 

All preorders will be subject to the above if not specifically noted that order is placed for RVC3 based product and someone would be willing to wait on RVC3 support which is at the moment TBD.

**Comment by szabi-luxonis** [comment]
> 5\. is that possible to provide a call-back API to allow user to define a function to fuse 3 dumpdispcost says (3,height,width,96) into 1 dumpdispcost (1,height,width,96) and pipe it back to compute the final depth and disparity? If this is possible, then the above API achitecture can be simplified to a single depth and disparity output instead of 3X each.

@ynjiun Feeding precomputed cost values to stereo HW is possible (however not exposed as API).
However dumping out 3x100 MB buffers puts a lot of stress on memory which slows down stereo/whole system. Other than that, fusing them takes a lot of time too, due to the sheer amount of data that needs processing. 
Not sure how cross-related-partial-dump helps, which you brought up in the mail, but feel free to expand the idea.

Cost values are refined using 3-way SGBM.
Dynamic programming of SGBM is the forward path of Viterbi Algorithm.
It computes the minimum accumulated cost of the path for all the disparity values of each pixel.
Because we don't know which disparity is finally taken, we need the cost for all possible disparity values.
Path for taking the minimum acuumulated cost can be totally different.

Also note that cost dump debug output is not available on RVC3. None of the debug outputs are available on RVC3. In addition confidence map output is not available either, since the HW doesn't support outputting it.
A simple way of fusing multiple disparity maps on RVC2 would be using the confidence map and taking the one that has the higher confidence (lower confidence score in our case).

**Comment by ynjiun** [comment]
> Cost values are refined using 3-way SGBM.
> Dynamic programming of SGBM is the forward path of Viterbi Algorithm.
> It computes the minimum accumulated cost of the path for all the disparity values of each pixel.
> Because we don't know which disparity is finally taken, we need the cost for all possible disparity values.
> Path for taking the minimum acuumulated cost can be totally different.

Hi @szabi-luxonis 

Thank you for the feedback.
Regarding "3-way SGBM", do you mean 3 disparities from right/left pair, right/middle pair and middle/left pair? Please elaborate. Thanks.

**Comment by szabi-luxonis** [comment]
> > Cost values are refined using 3-way SGBM.
> 
> Hi @szabi-luxonis
> 
> Thank you for the feedback.
> 
> Regarding "3-way SGBM", which 3-way? left-to-right (A1), top-left-to-bottom-right (A2), top-to-bottom (A3)? Thanks.

The SGBM algorithm paths implemented:
- left of the current pixel: →
- right of the current pixel: ← and
- vertically above the current pixel: ↓

**Comment by ynjiun** [comment]
> Also note that cost dump debug output is not available on RVC3. None of the debug outputs are available on RVC3. In addition confidence map output is not available either, since the HW doesn't support outputting it.

Hi @szabi-luxonis 

hmmm.... "confidence map output is not available either": is this also a h/w limitation? So what's supported in RVC3 then? could we have a list to compare RVC2 vs. RVC3 support outputs? Thank you.

**Comment by ynjiun** [comment]
> RVC3. In addition confidence map output is not available either, since the HW doesn't support outputting it.

Hi @szabi-luxonis 

For RVC3, would it be still possible to set the confidence threshold to remove those low confidence disparity?

Thanks.

**Comment by szabi-luxonis** [comment]
> > RVC3. In addition confidence map output is not available either, since the HW doesn't support outputting it.
> 
> Hi @szabi-luxonis
> 
> For RVC3, would it be still possible to set the confidence threshold to remove those low confidence disparity?
> 
> Thanks.

yes

**Comment by stephansturges** [comment]
Quick check-in: does this model come equipped with an IMU? 
I suspect I'm not the only one who is really interested in using this sensor for SLAM...

**Comment by cafemoloko** [comment]
@stephansturges OAK-D-LR comes with BMI270.

**Comment by Erol444** [comment]
Note: for production we plan to have BNO085/6. The background is that when we were designing the LR we didn't have any stock of BNO, but now that we have them we will redesign the device. Thoughts?

**Comment by stephansturges** [comment]
Thanks for confirming the early units come with BMI270. Is this the same unit as in the CM4-PoE?
BNV086 would be ideal from the specs that I see online... ultimately I think high-quality IMU will be very valuable for this sensor because the range makes it really useful for outdoor VIO / SLAM.

**Comment by stephansturges** [comment]
For some context: here is a great video from SpectacularAI on the importance of IMU quality for VSLAM https://www.youtube.com/watch?v=HpeVrCPTlF4

**Comment by Luxonis-David** [comment]
**Update:** We have shipped the first samples of OAK-D-LR yesterday so preorders will be fulfilled in these days just before the New Year. 
**Note:** This units may not have the best depth due to ongoing development, so recalibration later on might be needed to achieve best depth results. Since we do promote this product being in early access we figured out that it is better shipping the units sooner than later.

**Comment by ynjiun** [comment]
Hi @Luxonis-David 

Excited! Just received the unit.

Try to connect to USB and run a simple stereo depth. But encounter:
```
[1844301081575E1200] [1.11] [1.358] [ColorCamera(1)] [error] Camera not detected on socket: 2
[1844301081575E1200] [1.11] [1.359] [ColorCamera(0)] [error] Camera not detected on socket: 1
```
Could we post a simple demo python code to run the stereo depth for (Left/Right pair) and other pairs for example (Left/Middle pair) or (Middle/Right pair)?

Thanks a lot!

**Comment by themarpe** [comment]
@ynjiun 

Do you mind running on `develop` branch? There is also a depth_preview_lr.py on depthai-python repository

**Comment by ynjiun** [comment]
Sure. Could you provide me the link of develop branch of github? Do I need to reinstall the depthAI or just get depth_preview_lr.py to run on current main installation? thanks.

**Comment by themarpe** [comment]
@ynjiun 
Steps so far spot on, just run `python examples/install_requirements.py` and you should be set.

**Comment by ynjiun** [comment]
Hi @themarpe 

Cool. It works now! By the way, do I need to recalibrate the stereo cameras (rectified parameters)? if yes, please provide me the link (reference process) for re-calibration. Thanks a lot for your help!

**Comment by ynjiun** [comment]
> so recalibration later on might be needed to achieve best depth results.

Hi @Luxonis-David 

I got my unit running thru @themarpe's help. Now I tried to calibrate.py the unit and got the following error message:

`Error: Couldn't retrieve left, rigth and color frames for more than 5 seconds. Exiting...`

What shall I  update to proceed?

I did "git pull origin develop" where origin is https://github.com/luxonis/depthai.git

Please advise. Thanks.

**Comment by Luxonis-David** [comment]
@ynjiun we are still working on the depth quality so I wouldn't say that recalibrating the device at the moment would give any huge difference on what you already have in hand WRT depth quality. 
If you still want to recalibrate the device you can follow this steps: 

1. checkout `modular-calibration-lr` branch
2. Install requirements in `requirements.txt`
3. Run calibration script setting the parameters according to your target, example: 
`./calibrate.py --board resources/boards/OAK-D-LR.json -s 8.05 --defaultBoard -cd 0 -c 1 --fps 60`

Please note that you should use a bit bigger target for OAK-D-LR because of the larger baseline. A bigger TV would also do it for a start.

**Comment by themarpe** [comment]
@ynjiun 

You may calibrate OAK-D LR on `modular-calibration-lr` branch of `depthai`. We'll merge this in soon
Example of command line to do so: 
```
./calibrate.py --board resources/boards/OAK-D-LR.json -s 8.05 --defaultBoard -cd 0 -c 1
```

Note `-s 8.05` parameter should be changed according to your boards' square size

**Comment by ynjiun** [comment]
Hi @Luxonis-David and @themarpe 

the current calibration is off, there is almost no depth detected beyond 10 m.

I did 'git pull origin modular-calibration-lr" under depthai and run "pip install -r requirements.txt".
When I try to run "./calibrate.py --board resources/boards/OAK-D-LR.json -s 8.05 --defaultBoard -cd 0 -c 1"
I got follow errors:
```
~/oak/depthai$ ./calibrate.py --board resources/boards/OAK-D-LR.json -s 8.05 --defaultBoard -cd 0 -c 1
Traceback (most recent call last):
  File "./calibrate.py", line 1104, in <module>
    Main().run()
  File "./calibrate.py", line 324, in __init__
    self.device = dai.Device()
depthai.XLinkWriteError: Couldn't write data to stream: '__rpc_main' (X_LINK_ERROR)
```
What should I do next? Please advise. Thanks a lot.

**Comment by ynjiun** [comment]
Hi @Luxonis-David and @themarpe 

Thanks for the calibration code and it runs.
I use a 60" big screen TV for the chart with -s 8.25. After calibration, the rectification errors are all in range 0.1~0.2 and seems reasonably good (what's the typical expectation here? smaller than 0.1 to be considered as good calibration?). However, if I visually check the pixels, it might still off by about 2 pixels in y dimension. Plus when running the depth preview lr and I add the depth output, it seems the depth is off by a scale factor of 2/3(?).

Few questions:
1. in OAK-D-LR.json: hfov: 90; but according to [spec](https://shop.luxonis.com/collections/early-access/products/oak-d-lr-pre-order) the DFOV:96 => HFOV would be around 75 not 90 right?
2. OAK-D-LR native resolution is 1920x1200: what's the resolution used by calibration.py for calibration? resize to 1280x800 or crop to 1280x800?
3. when depth computation: depth = baseline*focal/disparity => what's the focal value in pixels? Is 1920x1200 and HFOV=90 used to compute focal? or 1280x800 and HFOV=?? used to compute focal?
4. Is there a way to calibrate using native resolution 1920x1200? if so the intrinsic/extrinsic matrix need to be changed for stereoDepth reification for 1280x800 resolution?

Please advise. Thank you very much for your help.

**Comment by Luxonis-David** [comment]
@ynjiun sorry about the delay, we used [M25360H06S](https://www.uctronics.com/m25360h06s.html) lens on these prototypes
1. According to my calculations the exact HFOV for the AR0234 sensor should be 91.7°.
![image](https://user-images.githubusercontent.com/75903226/210317935-37628dbe-f324-4c02-811b-7abe5d7ec727.png)
The calculations could be wrong so I will get the FOV measure on the samples today when I get to the office. 
For the rest I would need @themarpe to answer.

**Comment by themarpe** [comment]
@ynjiun 
2. OAK-D-LR gets calibrated at full 1920x1200 resolution
3. Focal is taken from intrinsics matrix first element if i'm not mistaken
4. Calibration happens at full while Stereo happens at downscaled 2/3 resolution. This change also modifies matrices accordingly. 

And thanks for the headsup on the Y displacement issue. Do you perhaps still have the dataset and any relevant logs of that? If so, do you mind sending those over for us to check

**Comment by Luxonis-David** [comment]
@ynjiun the above calculations are not correct as I suspected in first place, I have for some reason selected a Stereographic lens, but M25360H06S is rectilinear lens. 
The FOV is as follows and was also confirmed practically. 
Horiz. FoV --> 77.3 °
Vertical FoV --> 53.1 °
Diag. FoV --> 86.7 °
![image](https://user-images.githubusercontent.com/75903226/211256286-87c9b254-5b0d-4b9a-8b80-c987ca3faad9.png)
 
@chad-green the exact die size is 5805um x 3.645um thus the pixel count is  1935 x 1215. 
![image](https://user-images.githubusercontent.com/75903226/211256963-bee31409-bb7f-410d-81d8-7d6b2b74f607.png)

**Comment by stephansturges** [comment]
I’ve been working with the LR and for my use it’s already excellent, I’m excused to see how it improves with RVC3 and the final software.

quick question: is there any chance we will see full IP sealing on the final version? If not: can you develop a custom housing alternative or should I design my own? (this is crucial for my use case)

**Comment by Luxonis-David** [comment]
@stephansturges we are working on improving the IP rating of the device, especially the rain cover will be improved in order that we get better waterproofness than just from water splash. 
Second batch of the devices will already have this and you can expect those in March/April. 
X-coded M12 with waterproof USB connector is also on the roadmap but that would come later in the year. 
Thoughts?

**Comment by stephansturges** [comment]
@Luxonis-David are the first batch supposed to come with a rain cover already? I didn't receive any with my delivery at least,  I'd be curious to test this first and see if that will solve the use case for me.
To be honest I'm 90% certain that the rain cover won't cut it for a "final product" spec for my use case however, what I really need is the full waterproof spec of the CM4-PoE. 

Worst case I can design the casing myself as in the prototype I'm still seing overheating in continuous operation as will the CM4-PoE, even inside in 15 degrees C... so I might have to design my own case to solve the thermal issues anyway.

**Comment by stephansturges** [comment]
_@stephansturges we were not shipping the rain-covers because the design needs to be corrected first.
When we manage to get new covers we can ship them.
Will keep you updated._

-> gotcha... I misinterpreted your previous post as meaning there was a "gen1 raincover" that would be replaced with a "gen2 raincover" 

_About CM4, we did manufacture the updated design with USB HUB that does not overheat, you can request a replacement which should solve the issues with overheating. We are requesting shipping back the old units though._

That's very interesting... I have a number of those units that will need replacing, thanks for the tip.

**Comment by stephansturges** [comment]
FYI with the current design of the LR model I lose connection once the SHAVE cores get over 42 degrees celsius, which is about 1h30 of operation in the example below (in a small-ish room with ambient temperature at 18 degrees C). 

<img width="481" alt="image" src="https://user-images.githubusercontent.com/20320678/211633209-ea72bf5e-1604-4e72-9a26-28bd8e116340.png">

I'm going to test more and try to get more details, but I've observed this a dozen times already.

**Comment by themarpe** [comment]
@stephansturges thanks for the data point.
This is indeed an odd number, as it doesn't relate to anything (in terms of different processing, etc...). I would reckon that this is more of a time sensitive issue than temperature. 

If you have a MRE for this issue, we'll take a look and try to replicate on our end

**Comment by stephansturges** [comment]
> @stephansturges thanks for the data point. This is indeed an odd number, as it doesn't relate to anything (in terms of different processing, etc...). I would reckon that this is more of a time sensitive issue than temperature.
> 
> If you have a MRE for this issue, we'll take a look and try to replicate on our end

I think you are right that it's not a temperature problem, testing specifically for this issue I'm seeing it run stable up to 50 degC, and with the CM4 I only had systematic crashes when the SHAVE cores reached 60 degC...

I suspect this is something to do with the stereo algorithm because that's the only part of the process where I'm not catching errors, I'll experiment with that and let you know what I find.

**Comment by ynjiun** [comment]
> @ynjiun the above calculations are not correct as I suspected in first place, I have for some reason selected a Stereographic lens, but M25360H06S is rectilinear lens. The FOV is as follows and was also confirmed practically. Horiz. FoV --> 77.3 ° Vertical FoV --> 53.1 ° Diag. FoV --> 86.7 ° ![image](https://user-images.githubusercontent.com/75903226/211256286-87c9b254-5b0d-4b9a-8b80-c987ca3faad9.png)
> 
> @chad-green the exact die size is 5805um x 3.645um thus the pixel count is 1935 x 1215. ![image](https://user-images.githubusercontent.com/75903226/211256963-bee31409-bb7f-410d-81d8-7d6b2b74f607.png)

@Luxonis-David 

Thank you for the information. It helps a lot.

Question: I understand the current stereoDepth resolution limit is 1280x800. Is there a way to get a rectifiedLeft/Right with 1920x1080 resolution using existing depthai pipeline? Please refer to an example code to show how. Thanks for your help again.

**Comment by szabi-luxonis** [comment]
@stephansturges what is "SHAVE block temp?

**Comment by Luxonis-David** [comment]
@szabi-luxonis would you know the answer to the question that @ynjiun did post above about the depth resolution?

**Comment by szabi-luxonis** [comment]
@ynjiun the limit is 1280x65535 for stereo depth.
You can use ImageManip node and provide the rectification mesh to obtain rectified outputs at full resolution.

**Comment by stephansturges** [comment]
> @stephansturges what is "SHAVE block temp?

t.upa as reported by: 
https://docs.luxonis.com/projects/api/en/latest/samples/SystemLogger/system_information/

**Comment by stephansturges** [comment]
I think I've caught the error that is causing my pipeline to fail randomly. 
I don't have an MRE you can run yet but sharing the error in case it helps. 


<img width="1395" alt="image" src="https://user-images.githubusercontent.com/20320678/212071651-fba38e85-998c-43fd-bb3c-274fd069b980.png">

**Comment by ynjiun** [comment]
> You can use ImageManip or Warp node and provide the rectification mesh to obtain rectified outputs at full resolution.

@szabi-luxonis thank you for the information. Could you provide an example code or link to show how to retrieve the retification mesh from the firmware to setWarpMesh and activate ImageManip? Thanks a lot for your help.

**Comment by szabi-luxonis** [comment]
> > You can use ImageManip or Warp node and provide the rectification mesh to obtain rectified outputs at full resolution.
> 
> 
> 
> @szabi-luxonis thank you for the information. Could you provide an example code or link to show how to retrieve the retification mesh from the firmware to setWarpMesh and activate ImageManip? Thanks a lot for your help.

https://github.com/luxonis/depthai-python/blob/main/examples/ColorCamera/rgb_undistort.py

**Comment by chengguizi** [comment]
Is the hardware schematics available for OAK-D-LR? I could not find on https://github.com/luxonis/depthai-hardware (though I can find SR version already)

One thing I would like to check from the schematics is that:
- How is the three camera being synchronised now? Is it through FSIN? Or the old way that is using FW-based micro timing adjustment?

**Comment by Luxonis-David** [comment]
@chengguizi sensors can be synchronized through FSIN but that needs support that in FW, which will be done when the official release is made for OAK-D-LR if not earlier. 
I have attached the schematic of first design revision (R0M0E0), but please note that we are refactoring the schematic and also the layout will be updated with smaller corrections. 
Schematic: [Production.PDF](https://github.com/luxonis/depthai-hardware/files/10471881/Production.PDF)
 
@Luxonis-Blaz could you please upload the project files to the [hardware repo](https://github.com/luxonis/depthai-hardware) when you get a chance? Best would be to add files when the latest schematic with all corrections will be available and modified PCB layout so that it would serve as a good reference.

**Comment by chengguizi** [comment]
Thanks for the quick reply! @Luxonis-David 

I have just checked, it is great that all three cameras are connected together.

And yes, currently the FW in develop seems to have problem with hardware sync support. I tried putting camA as master (`dai::CameraControl::FrameSyncMode::OUTPUT`), and camB and camC as slave (`dai::CameraControl::FrameSyncMode::INPUT`). It didn't work.

If I turn off the hardware sync. I can see the camB and camC are effectively synchronised, but camA is like free running, not synced.

Is this expected, for now?

**Comment by chengguizi** [comment]
I also notice that the LED at the bottom changes color when i start streaming, pretty neat!

It seems to be achieved by IO38,39,43. Is this feature available for all OAK-SoM-Pro based product? Where do I find some documentation on the status of the LEDs. (I suppose it is a new feature:) )

![image](https://user-images.githubusercontent.com/7508112/213862675-35494fe6-026a-46fa-9757-d18e4a32e81c.png)

**Comment by chengguizi** [comment]
I have a question on what's happening when we feed color image stream to the StereoDepth Node.

https://github.com/luxonis/depthai-python/blob/develop/examples/StereoDepth/depth_preview_lr.py

From the example script, I can see the ColorCamera isp is directly fed into StereoDepth node, without converting to RAW8 first. Did the FW automatically add the color conversion itself? Or perhaps the StereoDepth conveniently just take the first channel, which i suppose is R channel?

**Comment by szabi-luxonis** [comment]
@chengguizi StereoDepth takes the first image plane. First image plane of NV12/YUV420 is the luminance (grey).

**Comment by chengguizi** [comment]
@szabi-luxonis Got it! That makes sense. So actually for all practical purpose, we can directly feed in color image to StereoDepth.  I vaguely remember this wasn't allow in the pass, on FW.

**Comment by szabi-luxonis** [comment]
@chengguizi it's been available for at least 1.5 years

**Comment by Luxonis-David** [comment]
@chengguizi LED is new feature not yet supported in full and future products will have the same LED functionality.
The existing products does not all have RGB LED nor the same controls. 
WRT syncing the cameras I think it is expected what you are experiencing, @themarpe would have better insight though.

**Comment by chengguizi** [comment]
@Luxonis-David , would love to know if we can start using the same design (the exact 3 IO ports) on custom baseboard that uses OAK-SoM-Pro? It seems like a neat feature to have this programmable status LED light.

**Comment by chengguizi** [comment]
@themarpe  Regarding the camera sync, is that the current known limitation that we can't get all three AR0234 hardware synced? Couldn't try fused depth map because of that :x

**Comment by Luxonis-David** [comment]
@chengguizi I would suggest that you use the following for LED control: 
Red      --> IO61
Green  --> IO59
Blue     --> IO60

Please note that those are GPIO-BOOT bootstrap pins, so only PU/PD that define the boot mode should be used for setting initial LED state. 

![image (139)](https://user-images.githubusercontent.com/75903226/214529460-5d7f2dff-6300-409a-8d20-ae0397c1163b.png)

**Comment by chengguizi** [comment]
@Luxonis-David  Got it, so you are suggesting we as the end user should use these three GPIOs instead, not what currently on the OAK-D-LR.

However I am a bit confused how could I use these three pins both as BOOT pins, and LED control. Wouldn't the addition of LEDs there destroy the logic of PU/PD resistors there?

**Comment by Luxonis-David** [comment]
Using MOSFET won't change the logic. 
![image](https://user-images.githubusercontent.com/75903226/214795207-88e048f4-a2d3-45ee-8628-3e1254d7e2ff.png)

**Comment by szabi-luxonis** [comment]
https://github.com/luxonis/depthai-python/pull/750

Adds support for stereo between any sensors for OAK D LR.

**Comment by ynjiun** [comment]
Any demo video or link on showing how to change lens for OAK D LR? please advise. Thank you.

**Comment by Luxonis-David** [comment]
@ynjiun we are preparing the docs for stacking and once available I will share here.

**Comment by chengguizi** [comment]
> [luxonis/depthai-python#750](https://github.com/luxonis/depthai-python/pull/750)
> 
> Adds support for stereo between any sensors for OAK D LR.

May I know how is the three camera synced in this example? I didn't see it being explicitly configured in the pipeline.

In my own test, CAMA seems free running.

**Comment by Luxonis-David** [comment]
@ynjiun here is the guide for removing the front lid and replacing the lens:
[OAK-D-LR_manual.pdf](https://github.com/luxonis/depthai-hardware/files/10555873/OAK-D-LR_manual.pdf)
Videos of disassembly are uploaded [here](https://drive.google.com/drive/folders/1V1msPKnKLGEeKTfczcrrr89GoC4hLSlW?usp=sharing).

**Comment by ynjiun** [comment]
> @ynjiun we are still working on the depth quality so I wouldn't say that recalibrating the device at the moment would give any huge difference on what you already have in hand WRT depth quality. If you still want to recalibrate the device you can follow this steps:
> 
> 1. checkout `modular-calibration-lr` branch
> 2. Install requirements in `requirements.txt`
> 3. Run calibration script setting the parameters according to your target, example:
>    `./calibrate.py --board resources/boards/OAK-D-LR.json -s 8.05 --defaultBoard -cd 0 -c 1 --fps 60`
> 
> Please note that you should use a bit bigger target for OAK-D-LR because of the larger baseline. A bigger TV would also do it for a start.

@Luxonis-David 

Thank you for the note. I have been trying couple times to use a 60" big screen TV for calibration but without success. In this case, instead of moving TV around, I have to move the OAK-D-LR for capturing those 13 images for calibration. But it's hard to really move the camera to those perspective view due to the counter intuitiveness of camera/target reverse relationship. (Moving TV will be more intuitive but it's impossible to do that without a mechanical setup). I wonder if you have any suggestion to perform OAK-D-LR calibration accurately, repeatably with ease. Please advise. Thanks a lot for your help in advance.

**Comment by Luxonis-David** [comment]
@ynjiun At the moment I don't really have anything that would ease the calibration besides that you can use the parameter `-ih` to invert/mirror the horizontal axis. 
We are also working on the TV calibration where you have three LCD screens being positioned orthogonally with the same origin. This will ease the calibration a lot, but I think we are not there yet so we could calibrate the OAK-D-LR with TV calibration approach. 
![image](https://user-images.githubusercontent.com/75903226/218972276-0612a259-dd7a-452a-8743-cbce22c04ece.png)
 CC: @themarpe

**Comment by chengguizi** [comment]
@themarpe Any updates on hardware sync all three AR0234 cameras on the LR units? Is it already supported, or still being worked on? Thanks!

**Comment by Erol444** [comment]
Hi @chengguizi , could you please create Issue/feature request on depthai-core? Just for tracking purposes for our development team.
Thanks, Erik

**Comment by themarpe** [comment]
@chengguizi currently available as a temporary change in `modular-calibration-lr` on depthai repo if you install requirements. But will soon land in `develop` of Core as well

**Comment by Erol444** [comment]
Perhaps it has already been merged to develop ([commit here](https://github.com/Luxonis/depthai-python/commit/82c0e7ee16d3be6444315689239e297e421066f4))?

**Comment by themarpe** [comment]
Correct, also in latest `develop` now:)

**Comment by chengguizi** [comment]
@Erol444 Ok let me see if it is indeed not done yet!
@themarpe Could you shine some light on what is the change to make it work? Will this change also applies to all other SoM based product, that carries all-AR0234 sensors? Or it is only OAK-D-LR specific, and Luxonis has a way to detect it is a OAK-D-LR?

**Comment by themarpe** [comment]
@chengguizi at the moment this is an OAK-D-LR specific. A common FSync between all sensors must be available, which then drives the sensors to capture at the same time. We are yet to expose this mechanism in a general way however

**Comment by themarpe** [comment]
@chengguizi 
Yes it'd be possible (yes it's driven by a VPU pin) - however generalizing this isn't currently on our agenda for rvc2 at the moment. Feel free to reachout to our email if commercially/business critical.

**Comment by stephansturges** [comment]
> 

Hey guys, is there any update on the IP-rated version of this sensor? I'm getting close to the point where I will be able to test this in real-world conditions and IP rating would make that a lot easier!

**Comment by Luxonis-David** [comment]
@stephansturges yes we did improve the IP rating by existing design adding rubber sealings, improving rain cover...
We are validating the design on which IP5X can we rate the enclosure with these days and next production batch will be made with the latest design improving IP rating.
Will keep you posted! 
 
![image (19)](https://user-images.githubusercontent.com/75903226/226533917-aef9af74-f4a6-48bb-9782-56761e83d3de.png)

**Comment by stephansturges** [comment]
Is there any 

> > @ynjiun we are still working on the depth quality so I wouldn't say that recalibrating the device at the moment would give any huge difference on what you already have in hand WRT depth quality. If you still want to recalibrate the device you can follow this steps:
> > 
> > 1. checkout `modular-calibration-lr` branch
> > 2. Install requirements in `requirements.txt`
> > 3. Run calibration script setting the parameters according to your target, example:
> >    `./calibrate.py --board resources/boards/OAK-D-LR.json -s 8.05 --defaultBoard -cd 0 -c 1 --fps 60`
> > 
> > Please note that you should use a bit bigger target for OAK-D-LR because of the larger baseline. A bigger TV would also do it for a start.

Is there are specific branch that I should use to try the latest improvements in depth quality for the LR sensor? 
I also have one module that is badly calibrated, so any tips / best practices for calibration would be great too.

**Comment by stephansturges** [comment]
Hey guys, is there any update on the IP5X rated-version, or future availability of the current version? I will probably need better IP rating and will have to make my own case anyway, but I need to start planning :)

**Comment by Erol444** [comment]
Hi @stephansturges , We actually have that in plan already - our next production batch OAK-D-LR will (should) be IP5X rated :)

**Comment by stephansturges** [comment]
@Erol444 yep I saw that but I think I will need IP63 at minimum so I will need to make my own enclosures probably, which is why I was curious on when the first samples from the final design will be available for me to work with 😄

**Comment by jingnanshi** [comment]
Any information on when will the camera be in stock? We would like to try this camera in our robotics lab for outdoor depths sensing. @Erol444

**Comment by Luxonis-David** [comment]
@GoldenZephyr unfortunately OAK-D-LR was not targeted for applications where time synchronization is needed, so you are right, it does not support the hardware FrameSync signal out of the box. 
Maybe you can use NTP software host time syncing protocol though, if your timings does not need to be really precise.
Thoughts?

**Comment by robsan7777** [comment]
Hello, are there any updates on the shipment. We order a couple of these and were just wondering when they would be sent.

**Comment by Erol444** [comment]
Hi @robsan7777 , I believe we plan to ship these in January.

**Comment by Erol444** [comment]
Hi @robsan7777 , is your order number 15425? If so, then we should start shipping these in a week or two, we are now working on the calibration of these devices. We apologize for the delay.

**Comment by robsan7777** [comment]
Hello,
I am trying to run the depthai_demo.py and for some reason I get a resolution error. Can I use this script with the new LR camera? The error is shown below:

[18443010E10D47F500] [1.12] [1.241] [ColorCamera(4)] [warning] Unsupported resolution set for detected camera AR0234, needs 1200_P, defaulting to it
[18443010E10D47F500] [1.12] [1.241] [MonoCamera(2)] [error] AR0234 only supports THE_1200_P resolution, defaulting to it
[18443010E10D47F500] [1.12] [1.241] [MonoCamera(0)] [error] AR0234 only supports THE_1200_P resolution, defaulting to it
RuntimeError: StereoDepth(7) - StereoDepth | Invalid resolution: Width 1920, Height 1200
Stopping demo...
=== TOTAL FPS ===
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "depthai_demo.py", line 336, in stop
    self._pv.closeQueues()
  File "/home/roberto/.local/lib/python3.8/site-packages/depthai_sdk/managers/preview_manager.py", line 100, in closeQueues
    for queue in self.outputQueues:
AttributeError: 'PreviewManager' object has no attribute 'outputQueues'

**Comment by jimas95** [comment]
@robsan7777 
running on the same issue, any solutions ? 

Also the installer for windows is not working for me

**Comment by robsan7777** [comment]
@jimas95 
I was hoping to have an SDK for the camera for easier integration but I do not think there is yet.

I tried multiple methods and ended up setting the ros_depthai_driver on ROS Iron since it is supported in just a few distributions. Was able to only see live camera images but not depth or anything else. It is a relatively new camera, not sure if they are still working on the drivers and supportive software for it.

**Comment by robsan7777** [comment]
@jimas95
Figured out the initial problem, when using the ethernet port from my laptop directly with the camera it does not find it sometimes. I used now a USB to Ethernet adapter and the depthai_demp works fine
